compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -473,173 +474,226 @@ object Implicits {
+          addPath(pre.cls.sourceModule.termRef)
+        case pre: TermRef if !isExcluded(pre.symbol) =>
+          pre.info match
+            case info: SingletonType =>
+              addPath(info)
+            case info: TypeRef if info.symbol.is(Module) =>
+              addCompanion(info.prefix, info.symbol.sourceModule)
+              addPath(info.prefix)
+            case _ =>
+              companions += pre
               addPath(pre.prefix)
-          case _ =>
-
-        tp.widenDealias match {
-          case tp: TypeRef =>
-            val sym = tp.symbol
-            if (isAnchor(sym)) {
-              val pre = tp.prefix
-              addPath(pre)
-              if (sym.isClass) addCompanion(pre, sym.companionModule)
-              else addCompanion(pre,
-                pre.member(sym.name.toTermName)
-                  .suchThat(companion => companion.is(Module) && companion.owner == sym.owner)
-                  .symbol)
-            }
-            val superAnchors = if (sym.isClass) tp.parents else anchors(tp.superType)
-            for (anchor <- superAnchors) comps ++= iscopeRefs(anchor)
-          case tp =>
-            for (part <- tp.namedPartsWith(_.isType)) comps ++= iscopeRefs(part)
-        }
-        comps
-      }
+        case _ =>
 
-   /** The implicit scope of type `tp`
-     *  @param isLifted    Type `tp` is the result of a `liftToAnchors` application
-     */
-    def iscope(tp: Type, isLifted: Boolean = false): OfTypeImplicits = {
-      val canCache = Config.cacheImplicitScopes && tp.hash != NotCached && !tp.isProvisional
-      def computeIScope() = {
-        val liftedTp = if (isLifted) tp else liftToAnchors(tp)
-        val refs =
-          if (liftedTp ne tp) {
-            implicitsDetailed.println(i"lifted of $tp = $liftedTp")
-            iscope(liftedTp, isLifted = true).companionRefs
-          }
-          else
-            collectCompanions(tp)
-        val result = new OfTypeImplicits(tp, refs)(runContext)
-        if (canCache &&
-            ((tp eq rootTp) ||          // first type traversed is always cached
-             !incomplete.contains(tp))) // other types are cached if they are not incomplete
-          implicitScopeCache(tp) = result
-        result
-      }
-      if (canCache) implicitScopeCache.getOrElse(tp, computeIScope())
-      else computeIScope()
-    }
+      parts.foreach(addCompanions)
+      companions
+    end collectCompanions
+
+    def recur(tp: Type): OfTypeImplicits =
+      val (parts, provisional) = collectParts(tp)
+      val companions = collectCompanions(tp, parts)
+      val result = OfTypeImplicits(tp, companions)(runContext)
+      if Config.cacheImplicitScopes
+         && tp.hash != NotCached
+        && !provisional
+        && (tp eq rootTp)              // first type traversed is always cached
+           || !incomplete.contains(tp) // other types are cached if they are not incomplete
```

Comment:
A pair of parentheses are missing here, which means this last condition is top-level, that is: all types that are not incomplete are cached, even if:
* Config.cacheImplicitScopes is false, or
* the types hash is NotCached, or
* the type is provisional